circuits/cpp/src/aztec3/msgpack/cbind_impl.hpp

```diff
@@ -0,0 +1,148 @@
+#pragma once
+// Note: heavy header due to serialization logic, don't include if msgpack.hpp will do
+// CBinding helpers that take a function or a lambda and
+// - bind the input as a coded msgpack array of all the arguments (using template metamagic)
+// - bind the return value to an out buffer, where the caller must free the memory
+#include <barretenberg/common/msgpack_impl.hpp>
+#include <aztec3/msgpack/schema_impl.hpp>
+#include <cstring>
+#include <type_traits>
+
+namespace msgpack {
+inline std::pair<uint8_t*, size_t> encode_buffer(auto&& obj)
+{
+    // Create a buffer to store the encoded data
+    msgpack::sbuffer buffer;
+    msgpack::pack(buffer, obj);
+
+    uint8_t* output = (uint8_t*)aligned_alloc(64, buffer.size());
+    memcpy(output, buffer.data(), buffer.size());
+    // Convert the buffer data to a string and return it
+    return { output, buffer.size() };
+}
+
+template <typename T> inline void decode(T* value, const uint8_t* encoded_data, size_t encoded_data_size)
+{
+    // Create a MsgPack unpacker
+    msgpack::unpack((const char*)encoded_data, encoded_data_size).get().convert(*value);
+}
+
+std::string string_encode(const auto& obj)
+{
+    std::stringstream output;
+    msgpack::pack(output, obj);
+    std::string output_str = output.str();
+    msgpack::object_handle oh = msgpack::unpack(output_str.data(), output_str.size());
+    std::stringstream pretty_output;
+    pretty_output << oh.get() << std::endl;
+    return pretty_output.str();
```

Comment:
This is essentially a JSON print in the end. It is necessary because msgpack::object is the only thing that can print to JSON, and it comes from an unpack